### PR TITLE
[Feat][Router]: add Llama Stack as a VectorStoreBackend for RAG

### LINKED ISSUE #1305: [Feat] Add Llama Stack as a VectorStoreBackend for RAG
## Summary

Add [Llama Stack](https://github.com/meta-llama/llama-stack) as a pluggable backend option for the Vector Store feature (#1262), enabling vSR to delegate document storage, embedding, and retrieval to a locally-running Llama Stack instance while keeping routing intelligence in vSR.

## Context

The Vector Stores API (#1262) is implemented with a `VectorStoreBackend` interface and two backends:
- `"memory"` — in-memory, pure Go, for dev/testing
- `"milvus"` — Milvus vector DB, for production

Llama Stack (Meta's open-source framework for Llama model applications) provides overlapping RAG capabilities — vector storage, embedding, chunking, retrieval — with support for multiple vector DBs (ChromaDB, Milvus, Qdrant, Weaviate, PGVector, SQLite-vec). It runs fully locally (no data leaves the network).

This issue proposes adding `"llama_stack"` as a third backend type. Initiated by collaboration discussion with the Llama Stack team.

## Why

- **Avoid duplicate infrastructure** — enterprises already running Llama Stack shouldn't need a separate vector DB setup for vSR
- **Leverage Llama Stack's mature storage layer** — battle-tested implementations for 6+ vector DBs
- **vSR keeps its unique value** — routing intelligence (which knowledge base to search, when to inject context, which model to route to) stays in vSR
- **Both run locally** — maintains the "no data leaves the network" promise

## Architecture

```
User query → Envoy → vSR (routing + RAG decision)
                        ↓
                   VectorStoreBackend interface
                   ├── "memory"       → in-memory (existing)
                   ├── "milvus"       → direct Milvus (existing)
                   └── "llama_stack"  → Llama Stack API (this issue)
                                           ↓
                                      Llama Stack (local, port 8321)
                                      ├── vector_io.insert()
                                      ├── vector_io.query()
                                      └── vector_dbs.register()
```

## Config

```yaml
vector_store:
  enabled: true
  backend_type: "llama_stack"
  llama_stack:
    endpoint: "http://localhost:8321"
    # auth_token: ""  # optional, if Llama Stack requires auth
```

### Files to create

**`src/semantic-router/pkg/vectorstore/llama_stack_backend.go`**

Implements the existing `VectorStoreBackend` interface (defined in `service.go`):

```go
type VectorStoreBackend interface {
    CreateCollection(ctx context.Context, vectorStoreID string, dimension int) error
    DeleteCollection(ctx context.Context, vectorStoreID string) error
    CollectionExists(ctx context.Context, vectorStoreID string) (bool, error)
    InsertChunks(ctx context.Context, vectorStoreID string, chunks []EmbeddedChunk) error
    DeleteByFileID(ctx context.Context, vectorStoreID string, fileID string) error
    Search(ctx context.Context, vectorStoreID string, queryEmbedding []float32,
           topK int, threshold float32, filter map[string]interface{}) ([]SearchResult, error)
    Close() error
}
```

Map each method to Llama Stack's API:

| VectorStoreBackend method | Llama Stack API call |
|---|---|
| `CreateCollection(vsID, dim)` | `POST /vector-dbs` — `vector_dbs.register(vector_db_id, embedding_model)` |
| `DeleteCollection(vsID)` | `DELETE /vector-dbs/{id}` — `vector_dbs.unregister(vector_db_id)` |
| `CollectionExists(vsID)` | `GET /vector-dbs/{id}` — check if registered |
| `InsertChunks(vsID, chunks)` | `POST /vector-io/insert` — `vector_io.insert(vector_db_id, chunks)` |
| `DeleteByFileID(vsID, fileID)` | May need to track chunk IDs and delete individually |
| `Search(vsID, embedding, topK, ...)` | `POST /vector-io/query` — `vector_io.query(vector_db_id, query)` |

**Note on embeddings:** Llama Stack can handle embeddings internally. Two approaches:
1. **Let Llama Stack embed** — send raw text via `rag_tool.insert()`, Llama Stack chunks + embeds. Simpler but less control.
2. **vSR embeds, Llama Stack stores** — use `vector_io.insert()` with pre-computed embeddings. More control, keeps vSR's CandleEmbedder in the loop.

Recommend option 1 for simplicity — if using Llama Stack, use its full pipeline.

**`src/semantic-router/pkg/vectorstore/llama_stack_backend_test.go`**

- Unit tests with a mock HTTP server (httptest)
- Integration tests gated behind `SKIP_LLAMA_STACK_TESTS` env var (similar to Milvus pattern)

### Files to modify

**`src/semantic-router/pkg/vectorstore/factory.go`** — add case:
```go
case "llama_stack":
    return NewLlamaStackBackend(llamaStackCfg)
```

**`src/semantic-router/pkg/config/vectorstore.go`** — add config:
```go
type LlamaStackConfig struct {
    Endpoint  string `json:"endpoint" yaml:"endpoint"`
    AuthToken string `json:"auth_token,omitempty" yaml:"auth_token,omitempty"`
}
```
Add `LlamaStack *LlamaStackConfig` field to `VectorStoreConfig`.

**`src/semantic-router/cmd/main.go`** — add initialization case for llama_stack backend in the vector store init block (~line 305).

### Llama Stack API Reference

Llama Stack's vector/RAG APIs (from [docs](https://github.com/meta-llama/llama-stack/blob/main/docs/source/building_applications/rag.md)):

**Low-level vector_io:**
```python
# Register a vector DB
client.vector_dbs.register(vector_db_id="my_db", embedding_model="all-MiniLM-L6-v2")

# Insert chunks
client.vector_io.insert(vector_db_id="my_db", chunks=[
    {"content": "chunk text", "mime_type": "text/plain", "metadata": {"file_id": "..."}}
])

# Query
results = client.vector_io.query(vector_db_id="my_db", query="search text")
```

**High-level rag_tool (handles chunking + embedding):**
```python
# Insert documents (auto-chunks)
client.tool_runtime.rag_tool.insert(
    documents=[{"content": "full document text", "mime_type": "text/plain"}],
    vector_db_id="my_db",
    chunk_size_in_tokens=512
)

# Query
results = client.tool_runtime.rag_tool.query(
    vector_db_ids=["my_db"],
    content="search query"
)
```

### Prerequisites

- Llama Stack installed locally: `pip install llama-stack`
- Running: `llama stack run local --port 8321`
- A vector DB provider configured in Llama Stack (ChromaDB is the simplest for dev)

### Testing

1. Install Llama Stack and start it locally
2. Configure vSR with `backend_type: "llama_stack"` and `endpoint: "http://localhost:8321"`
3. Run the same smoke test that works with memory/milvus backends:
   - Create store → Upload file → Attach → Poll → Search → Verify results

## Dependencies

- #1262 (Vector Stores API) — must be merged first
- Llama Stack running locally for testing

## References

- [Llama Stack GitHub](https://github.com/meta-llama/llama-stack)
- [Llama Stack RAG docs](https://github.com/meta-llama/llama-stack/blob/main/docs/source/building_applications/rag.md)
- [Red Hat + Llama Stack RAG tutorial](https://developers.redhat.com/articles/2025/08/05/retrieval-augmented-generation-llama-stack-and-python)
- [Llama Stack detailed tutorial](https://llamastack.github.io/docs/getting_started/detailed_tutorial)
- Collaboration initiated via Slack with Francisco (Llama Stack maintainer)

## Estimated Scope

~200-300 lines of new code (backend + config + factory case). No changes to existing backend implementations. Follows the exact same pattern as the Milvus backend.

### PR DESCRIPTION
**Pull request by noalimoy**
### Summary

Add Llama Stack as a new `VectorStoreBackend` for the RAG pipeline, enabling document ingestion and semantic search through Meta's Llama Stack REST API.

This PR delivers the full integration across four layers:
- **Backend implementation** with unit and integration tests
- **Makefile targets** for local Llama Stack setup and testing
- **E2E profile** for Kubernetes-based validation in Kind clusters
- **Documentation** updates for user-facing configuration

### Motivation

Llama Stack provides a lightweight, self-hosted vector store backend that is OpenAI API-compatible, making it an accessible option for RAG workloads without requiring external services like Milvus. Enterprises already running Llama Stack can reuse their existing infrastructure instead of deploying a separate vector DB. This was requested in #1305 to expand the range of supported vector store backends.

### API Approach and Design Decisions

Following guidance from [Francisco](https://github.com/franciscojavierarceo) (Llama Stack maintainer), we use the **OpenAI-compatible API** instead of the `rag_tool`. However, Llama Stack's API and vSR's `VectorStoreBackend` interface differ in how they handle embeddings and resource identity, which required three key design choices:

**1. Search: text query vs. embedding vector**

The `VectorStoreBackend.Search()` interface passes a `queryEmbedding []float32` (a pre-computed embedding vector), but Llama Stack's search endpoint (`POST /v1/vector_stores/{id}/search`) expects a **text query string** -- it handles embedding internally. This is a fundamental mismatch.

**Solution:** The original query text is passed through `filter["_query_text"]` from the caller. Both `route_vectorstore.go` (REST API handler) and `req_filter_rag_vectorstore.go` (ExtProc filter) inject the query text into the filter map before calling `Search()`. The Llama Stack backend reads this text and sends it to the search endpoint, ignoring the embedding vector. Other backends (memory, milvus) safely ignore this extra filter key.

**2. InsertChunks: pre-chunked data vs. file upload**

the OpenAI-compatible file upload endpoint (`POST /v1/vector_stores/{id}/files`), accepts a raw file and handles chunking + embedding internally. But vSR's ingestion pipeline already chunks and embeds documents before calling `InsertChunks()`, passing `[]EmbeddedChunk` objects (text content + pre-computed embeddings).

**Solution:** For InsertChunks, we use the lower-level `POST /v1/vector-io/insert` endpoint, which accepts individual chunks with their text content. The pre-computed embeddings from vSR are not sent -- Llama Stack re-embeds using its own configured model (`sentence-transformers/all-MiniLM-L6-v2`), ensuring consistency between insert and search embeddings.

**3. ID mapping: vSR names vs. Llama Stack IDs**

vSR and Llama Stack each generate their own IDs for the same vector store (e.g., vSR: `vs_6a37f067-...`, Llama Stack: `vs_21c1db0f-...`). The backend keeps a `storeIDs` map to translate between the two, populated at creation time and refreshed via `resolveStoreID()` if needed.

**API mapping summary:**

| VectorStoreBackend method | Llama Stack endpoint | API layer |
|---------------------------|---------------------|-----------|
| `CreateCollection` | `POST /v1/vector_stores` | OpenAI-compatible |
| `DeleteCollection` | `DELETE /v1/vector_stores/{id}` | OpenAI-compatible |
| `CollectionExists` | `GET /v1/vector_stores/{id}` | OpenAI-compatible |
| `Search` | `POST /v1/vector_stores/{id}/search` | OpenAI-compatible |
| `InsertChunks` | `POST /v1/vector-io/insert` | Low-level vector-io |
| `DeleteByFileID` | In-memory tracking + re-insert | Application-level |

#### Phase 1: Backend Implementation

| File | Change |
|------|--------|
| `pkg/vectorstore/llama_stack_backend.go` | **New** - Full `VectorStoreBackend` implementation |
| `pkg/vectorstore/factory.go` | Register `"llama_stack"` backend type in factory |
| `pkg/vectorstore/types.go` | Add `BackendTypeLlamaStack` constant |
| `pkg/config/vectorstore.go` | Add `LlamaStackConfig` struct with endpoint, model, dimension |
| `cmd/main.go` | Wire Llama Stack config into server initialization |
| `pkg/apiserver/route_vectorstore.go` | Inject `_query_text` into filter map before Search |
| `pkg/extproc/req_filter_rag_vectorstore.go` | Inject `_query_text` into filter map before Search |
| `config/config-rag-example.yaml` | Add Llama Stack example configuration |

#### Phase 2: Tests

| File | Change |
|------|--------|
| `pkg/vectorstore/llama_stack_backend_test.go` | **New** - Unit tests (httptest mock) + integration tests |
| `pkg/vectorstore/memory_backend_test.go` | Add shared test helpers |

**Test coverage:**
- **Unit tests** using `httptest` mock server covering all CRUD operations, error handling, edge cases, timeout behavior, and concurrent access
- **Integration tests** against a live Llama Stack instance covering the full lifecycle: create store -> insert chunks -> search -> delete
- Integration tests gated behind `SKIP_LLAMA_STACK_TESTS` env var (default: `true`)

#### Phase 3: Makefile Integration

| File | Change |
|------|--------|
| `tools/make/llama-stack.mk` | **New** - Targets for Llama Stack setup and testing |
| `tools/make/build-run-test.mk` | Add `SKIP_LLAMA_STACK_TESTS` (default: `true`) to `test-semantic-router` |
| `Makefile` | Include `llama-stack.mk` |

**Available targets:**
```
make start-llama-stack             # Start Llama Stack + auto-register embedding model
make stop-llama-stack              # Stop Llama Stack
make restart-llama-stack           # Restart Llama Stack
make status-llama-stack            # Show status
make clean-llama-stack             # Remove Llama Stack
make test-llama-stack-vectorstore  # Run integration tests against Docker-based Llama Stack
```

#### Phase 4: E2E Profile & Documentation

| File | Change |
|------|--------|
| `e2e/profiles/rag-llama-stack/profile.go` | **New** - E2E profile implementation |
| `e2e/profiles/rag-llama-stack/values.yaml` | **New** - Helm values configuring `backend_type: "llama_stack"` with in-cluster endpoint |
| `e2e/profiles/rag-llama-stack/manifests/llama-stack.yaml` | **New** - K8s Namespace + Deployment + Service |
| `e2e/cmd/e2e/main.go` | Register `rag-llama-stack` profile |
| `tools/make/e2e.mk` | Add `e2e-test-rag-llama-stack` convenience target |
| `website/docs/installation/configuration.md` | Document `vector_store` with `llama_stack` option |

**Runs test case:** `rag-vectorstore` (backend-agnostic, validates full RAG pipeline)

### How to Test

**Unit tests (no infrastructure needed):**
```bash
make test-semantic-router
```

**Integration tests (requires Docker):**
```bash
make test-llama-stack-vectorstore
```

**E2E tests (requires Kind + Docker):**
```bash
make e2e-test-rag-llama-stack
```

### Configuration Example

```yaml
vector_store:
  enabled: true
  backend_type: "llama_stack"
  file_storage_dir: "/tmp/vsr-data"
  embedding_model: "mmbert"
  embedding_dimension: 384
  llama_stack:
    endpoint: "http://localhost:8321"
    embedding_model: "sentence-transformers/all-MiniLM-L6-v2"
```

### Test Plan

- [x] `go build ./...` passes
- [x] `go vet ./...` passes
- [x] Unit tests pass (`make test-semantic-router`)
- [x] Linter checks pass (errcheck, gofumpt, staticcheck)
- [x] Integration tests pass (`make test-llama-stack-vectorstore`)
- [x] E2E tests pass (`make e2e-test-rag-llama-stack`)

### Dependencies

- **Llama Stack image:** `llamastack/distribution-starter:0.5.0`
- **Embedding model:** `sentence-transformers/all-MiniLM-L6-v2` (384 dimensions)
- No new Go module dependencies (`go.mod` unchanged) -- the backend uses Go's standard `net/http` for all REST calls to Llama Stack, avoiding an external SDK

Resolves #1305

### CODE REVIEW

**Comment by noalimoy** [comment]
Hi [franciscojavierarceo](https://github.com/franciscojavierarceo) @yossiovadia 
Could you please review when you have a chance?

File: e2e/profiles/rag-llama-stack/profile.go [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+	timeoutDeploymentWait        = 30 * time.Minute
+	timeoutLlamaStackReady       = 10 * time.Minute
+	intervalServiceRetry         = 5 * time.Second
+	timeoutHTTPRequest           = 60 * time.Second
+	delayPortForwardReady        = 2 * time.Second
+
+	// Image constants
+	imageRepository = "ghcr.io/vllm-project/semantic-router/extproc"
+	imagePullPolicy = "Never"
+	llamaStackImage = "llamastack/distribution-starter:0.5.0"
```
Comment (franciscojavierarceo) [inline e2e/profiles/rag-llama-stack/profile.go:41]: nice, starter is a good distro.

**Comment by rootfs** [comment]
Also see the progress at vllm: https://github.com/vllm-project/vllm/pull/34487

**Comment by szedan-rh** [comment]
LGTM - Thanks Noa.

@rootfs - I think we should also update the documentation and mention that we have Llama stack backend supported so it will not be missed.

@rootfs / @franciscojavierarceo  - If no comments from you side also, will approve and merge the PR

**Review by franciscojavierarceo — APPROVED** [review]
yeah this lgtm, the only nit is that i'd recommend using milvus with the starter distro as we support hybrid search OOTB with it and right now i think the default in the distro image is just vector search.
